File: backend/src/main/java/com/paymentgateway/payment_gateway/controller/PaymentController.java
package com.paymentgateway.payment_gateway.controller;

import com.paymentgateway.payment_gateway.dto.ErrorResponse;
import com.paymentgateway.payment_gateway.dto.PaymentRequest;
import com.paymentgateway.payment_gateway.entity.Merchant;
import com.paymentgateway.payment_gateway.entity.Order;
import com.paymentgateway.payment_gateway.entity.Payment;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.paymentgateway.payment_gateway.entity.IdempotencyKey;
import com.paymentgateway.payment_gateway.jobs.ProcessPaymentJob;
import com.paymentgateway.payment_gateway.queue.JobQueue;
import com.paymentgateway.payment_gateway.repository.IdempotencyKeyRepository;
import com.paymentgateway.payment_gateway.service.MerchantService;
import com.paymentgateway.payment_gateway.service.OrderService;
import com.paymentgateway.payment_gateway.service.PaymentService;
import com.paymentgateway.payment_gateway.util.PaymentValidator;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/v1/payments")
@RequiredArgsConstructor
public class PaymentController {

    private final PaymentService paymentService;
    private final OrderService orderService;
    private final MerchantService merchantService;
    private final IdempotencyKeyRepository idempotencyKeyRepository;
    private final JobQueue jobQueue;
    private final ObjectMapper objectMapper;
    
    // Create Payment Endpoint
    @PostMapping
    public ResponseEntity<?> createPayment(
            @RequestHeader("X-Api-Key") String apiKey,
            @RequestHeader("X-Api-Secret") String apiSecret,
            @RequestHeader(value = "Idempotency-Key", required = false) String idempotencyKey,
            @RequestBody PaymentRequest request) {
        
        try {
            // Validate API credentials
            Optional<Merchant> merchantOpt = merchantService.findByApiKey(apiKey);
            if (merchantOpt.isEmpty() || !merchantOpt.get().getApiSecret().equals(apiSecret)) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                        .body(new ErrorResponse("AUTHENTICATION_ERROR", "Invalid API credentials"));
            }

            Merchant merchant = merchantOpt.get();

            // Idempotency Check
            if (idempotencyKey != null && !idempotencyKey.trim().isEmpty()) {
                Optional<IdempotencyKey> existingKeyOpt = idempotencyKeyRepository.findByKeyAndMerchantId(idempotencyKey, java.util.UUID.fromString(merchant.getId()));
                
                if (existingKeyOpt.isPresent()) {
                    IdempotencyKey existingKey = existingKeyOpt.get();
                    if (existingKey.getExpiresAt().isAfter(LocalDateTime.now())) {
                        // Return cached response
                        Object cachedResponse = objectMapper.readValue(existingKey.getResponse(), Map.class);
                        return ResponseEntity.status(HttpStatus.CREATED).body(cachedResponse);
                    } else {
                        // Expired - delete and process as new
                        idempotencyKeyRepository.delete(existingKey);
                    }
                }
            }

            // Validate order exists
            Optional<Order> orderOpt = orderService.findById(request.getOrderId());
            if (orderOpt.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(new ErrorResponse("NOT_FOUND_ERROR", "Order not found"));
            }

            Order order = orderOpt.get();

            // Verify order belongs to merchant
            if (!order.getMerchant().getId().equals(merchant.getId())) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                        .body(new ErrorResponse("BAD_REQUEST_ERROR", "Order does not belong to merchant"));
            }

            // Check if payment already exists for this order
            List<Payment> existingPayments = paymentService.findByOrderId(order.getId());
            if (!existingPayments.isEmpty()) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                        .body(new ErrorResponse("PAYMENT_ALREADY_EXISTS", "Payment already exists for this order"));
            }
            // Validate payment method
            String method = request.getMethod();
            if (!"upi".equals(method) && !"card".equals(method)) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                        .body(new ErrorResponse("BAD_REQUEST_ERROR", "Invalid payment method"));
            }

            // Validate method-specific fields
            String vpa = null;
            String cardNetwork = null;
            String cardLast4 = null;

            if ("upi".equals(method)) {
                // Validate VPA
                vpa = request.getVpa();
                if (!PaymentValidator.isValidVPA(vpa)) {
                    return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                            .body(new ErrorResponse("INVALID_VPA", "Invalid VPA format"));
                }
            } else if ("card".equals(method)) {
                // Validate card number
                String cardNumber = request.getCardNumber();
                if (!PaymentValidator.isValidCardNumber(cardNumber)) {
                    return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                            .body(new ErrorResponse("INVALID_CARD", "Invalid card number"));
                }

                // Detect card network
                cardNetwork = PaymentValidator.detectCardNetwork(cardNumber);

                // Validate expiry
                if (!PaymentValidator.isValidExpiry(request.getExpiryMonth(), request.getExpiryYear())) {
                    return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                            .body(new ErrorResponse("EXPIRED_CARD", "Card has expired or invalid expiry date"));
                }

                // Validate CVV
                if (!PaymentValidator.isValidCVV(request.getCvv(), cardNetwork)) {
                    return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                            .body(new ErrorResponse("INVALID_CARD", "Invalid CVV"));
                }

                // Store only last 4 digits
                cardLast4 = PaymentValidator.getCardLast4(cardNumber);
            }

            // Create payment
            Payment payment = paymentService.createPayment(
                    order, merchant, method, vpa, cardNetwork, cardLast4
            );

            // Enqueue Job for Async Processing
            jobQueue.enqueuePaymentJob(payment.getId());

            // Build response
            Map<String, Object> response = new HashMap<>();
            response.put("id", payment.getId());
            response.put("order_id", order.getId());
            response.put("merchant_id", merchant.getId());
            response.put("amount", payment.getAmount());
            response.put("currency", payment.getCurrency());
            response.put("method", payment.getMethod());
            response.put("status", payment.getStatus()); // Should be "pending"
            
            if ("upi".equals(method)) {
                response.put("vpa", payment.getVpa());
            } else if ("card".equals(method)) {
                response.put("card_network", payment.getCardNetwork());
                response.put("card_last4", payment.getCardLast4());
            }
            
            response.put("created_at", payment.getCreatedAt().toString());
            // response.put("updated_at", payment.getUpdatedAt().toString()); // Updated At might be null initially or same as created

            // Store idempotency key if provided
            if (idempotencyKey != null && !idempotencyKey.trim().isEmpty()) {
                IdempotencyKey keyRecord = new IdempotencyKey();
                keyRecord.setKey(idempotencyKey);
                keyRecord.setMerchantId(java.util.UUID.fromString(merchant.getId()));
                keyRecord.setResponse(objectMapper.writeValueAsString(response));
                idempotencyKeyRepository.save(keyRecord);
            }

            return ResponseEntity.status(HttpStatus.CREATED).body(response);

        } catch (IllegalArgumentException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(new ErrorResponse("BAD_REQUEST_ERROR", e.getMessage()));
        } catch (Exception e) {
            e.printStackTrace(); // Log error
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ErrorResponse("INTERNAL_ERROR", "An unexpected error occurred"));
        }
    }

    // Get Payment Endpoint
    @GetMapping("/{paymentId}")
    public ResponseEntity<?> getPayment(
            @RequestHeader("X-Api-Key") String apiKey,
            @RequestHeader("X-Api-Secret") String apiSecret,
            @PathVariable String paymentId) {
        
        try {
            // Validate API credentials
            Optional<Merchant> merchantOpt = merchantService.findByApiKey(apiKey);
            if (merchantOpt.isEmpty() || !merchantOpt.get().getApiSecret().equals(apiSecret)) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                        .body(new ErrorResponse("AUTHENTICATION_ERROR", "Invalid API credentials"));
            }

            Merchant merchant = merchantOpt.get();

            // Find payment
            Optional<Payment> paymentOpt = paymentService.findById(paymentId);
            if (paymentOpt.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(new ErrorResponse("NOT_FOUND_ERROR", "Payment not found"));
            }

            Payment payment = paymentOpt.get();

            // Verify payment belongs to merchant
            if (!payment.getMerchant().getId().equals(merchant.getId())) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(new ErrorResponse("NOT_FOUND_ERROR", "Payment not found"));
            }

            // Build response
            Map<String, Object> response = new HashMap<>();
            response.put("id", payment.getId());
            response.put("order_id", payment.getOrder().getId());
            response.put("merchant_id", merchant.getId());
            response.put("amount", payment.getAmount());
            response.put("currency", payment.getCurrency());
            response.put("method", payment.getMethod());
            response.put("status", payment.getStatus());
            
            if ("upi".equals(payment.getMethod())) {
                response.put("vpa", payment.getVpa());
            } else if ("card".equals(payment.getMethod())) {
                response.put("card_network", payment.getCardNetwork());
                response.put("card_last4", payment.getCardLast4());
            }
            
            if (payment.getErrorCode() != null) {
                response.put("error_code", payment.getErrorCode());
                response.put("error_description", payment.getErrorDescription());
            }
            
            response.put("created_at", payment.getCreatedAt().toString());
            response.put("updated_at", payment.getUpdatedAt().toString());

            return ResponseEntity.ok(response);

        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ErrorResponse("INTERNAL_ERROR", "An unexpected error occurred"));
        }
    }

    // Capture Payment Endpoint
    @PostMapping("/{paymentId}/capture")
    public ResponseEntity<?> capturePayment(
            @RequestHeader("X-Api-Key") String apiKey,
            @RequestHeader("X-Api-Secret") String apiSecret,
            @PathVariable String paymentId,
            @RequestBody com.paymentgateway.payment_gateway.dto.CaptureRequest request) {

        try {
            // Validate API credentials
            Optional<Merchant> merchantOpt = merchantService.findByApiKey(apiKey);
            if (merchantOpt.isEmpty() || !merchantOpt.get().getApiSecret().equals(apiSecret)) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                        .body(new ErrorResponse("AUTHENTICATION_ERROR", "Invalid API credentials"));
            }
            
            Merchant merchant = merchantOpt.get();
            
            // Check if payment belongs to merchant is handled implicitly by service if we passed merchantId, 
            // but here we fetch by ID. Ideally we should verify ownership.
            Optional<Payment> paymentOpt = paymentService.findById(paymentId);
            if (paymentOpt.isPresent() && !paymentOpt.get().getMerchant().getId().equals(merchant.getId())) {
                 return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(new ErrorResponse("NOT_FOUND_ERROR", "Payment not found"));
            }

            Payment payment = paymentService.capturePayment(paymentId, request.getAmount());

            // Build response
            Map<String, Object> response = new HashMap<>();
            response.put("id", payment.getId());
            response.put("order_id", payment.getOrder().getId());
            // response.put("merchant_id", merchant.getId()); // Optional in response
            response.put("amount", payment.getAmount());
            response.put("currency", payment.getCurrency());
            response.put("method", payment.getMethod());
            response.put("status", payment.getStatus());
            response.put("captured", payment.getCaptured());
            
            response.put("created_at", payment.getCreatedAt().toString());
            response.put("updated_at", payment.getUpdatedAt().toString());

            return ResponseEntity.ok(response);

        } catch (IllegalArgumentException e) {
             return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(new ErrorResponse("BAD_REQUEST_ERROR", e.getMessage())); // "Payment not in capturable state" maps here
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ErrorResponse("INTERNAL_ERROR", "An unexpected error occurred"));
        }
    }
    // List All Payments Endpoint (for dashboard)
    @GetMapping
    public ResponseEntity<?> listPayments(
            @RequestHeader("X-Api-Key") String apiKey,
            @RequestHeader("X-Api-Secret") String apiSecret) {
        
        try {
            // Validate API credentials
            Optional<Merchant> merchantOpt = merchantService.findByApiKey(apiKey);
            if (merchantOpt.isEmpty() || !merchantOpt.get().getApiSecret().equals(apiSecret)) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                        .body(new ErrorResponse("AUTHENTICATION_ERROR", "Invalid API credentials"));
            }

            Merchant merchant = merchantOpt.get();

            // Get all payments for this merchant
            List<Payment> payments = paymentService.findByMerchantId(merchant.getId());

            // Build response list
            List<Map<String, Object>> responseList = new ArrayList<>();
            
            for (Payment payment : payments) {
                Map<String, Object> paymentData = new HashMap<>();
                paymentData.put("id", payment.getId());
                paymentData.put("order_id", payment.getOrder().getId());
                paymentData.put("merchant_id", merchant.getId());
                paymentData.put("amount", payment.getAmount());
                paymentData.put("currency", payment.getCurrency());
                paymentData.put("method", payment.getMethod());
                paymentData.put("status", payment.getStatus());
                
                if ("upi".equals(payment.getMethod())) {
                    paymentData.put("vpa", payment.getVpa());
                } else if ("card".equals(payment.getMethod())) {
                    paymentData.put("card_network", payment.getCardNetwork());
                    paymentData.put("card_last4", payment.getCardLast4());
                }
                
                if (payment.getErrorCode() != null) {
                    paymentData.put("error_code", payment.getErrorCode());
                    paymentData.put("error_description", payment.getErrorDescription());
                }
                
                paymentData.put("created_at", payment.getCreatedAt().toString());
                paymentData.put("updated_at", payment.getUpdatedAt().toString());
                
                responseList.add(paymentData);
            }

            return ResponseEntity.ok(responseList);

        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(new ErrorResponse("INTERNAL_ERROR", "An unexpected error occurred"));
        }
    }
}

File: backend/src/controllers/paymentController.ts
import { Request, Response } from 'express';
import { query } from '../config/db';
import { v4 as uuidv4 } from 'uuid';
import { paymentQueue, webhookQueue } from '../config/queue';
import { validateVPA, validateLuhn, detectCardNetwork, validateExpiry } from '../utils/validation';

const generatePaymentId = () => {
    const randomChars = uuidv4().replace(/-/g, '').substring(0, 16);
    return `pay_${randomChars}`;
};

export const createPayment = async (req: Request, res: Response) => {
    const body = req.body;
    const merchant = req.merchant;
    const idempotencyKey = req.headers['idempotency-key'] as string;

    const order_id = body.order_id;
    const method = body.method;
    const vpa = method == "upi" ? body.vpa : "";
    const card = method == "card" ? body.card : {};
    try {
        // 0. Idempotency Check
        if (idempotencyKey) {
            const idempotencyResult = await query(
                'SELECT * FROM idempotency_keys WHERE key = $1 AND merchant_id = $2',
                [idempotencyKey, merchant.id]
            );

            if (idempotencyResult.rows.length > 0) {
                const record = idempotencyResult.rows[0];
                const now = new Date();

                if (new Date(record.expires_at) > now) {
                    // console.log(`[Idempotency] returning cached response for key: ${idempotencyKey}`);
                    return res.status(201).json(record.response);
                } else {
                    // Expired, delete and continue
                    await query('DELETE FROM idempotency_keys WHERE key = $1 AND merchant_id = $2', [idempotencyKey, merchant.id]);
                }
            }
        }

        // 1. Verify Order
        let order;
        try {
            const orderResult = await query('SELECT * FROM orders WHERE id = $1 AND merchant_id = $2', [order_id, merchant.id]);

            if (orderResult.rows.length === 0) {
                return res.status(404).json({
                    error: {
                        code: 'NOT_FOUND_ERROR',
                        description: 'Order not found or does not belong to merchant'
                    }
                });
            }
            order = orderResult.rows[0];

            // 1.1 Check for existing payments for this order
            // Requirement: Block creation if payment is 'pending' or 'success'. Allow if 'failed'.
            const existingPaymentResult = await query(
                `SELECT id, status FROM payments 
                 WHERE order_id = $1 
                 AND status IN ('pending', 'processing', 'success')`,
                [order_id]
            );

            if (existingPaymentResult.rows.length > 0) {
                return res.status(400).json({
                    error: {
                        code: 'BAD_REQUEST_ERROR',
                        description: `Order already has a payment in ${existingPaymentResult.rows[0].status} status`
                    }
                });
            }

        } catch (error) {
            return res.status(500).json({ error: { code: 'INTERNAL_SERVER_ERROR', description: 'DB Error' } });
        }

        // 2. method-specific validation & field extraction
        let methodSpecificFields = {};
        let cardNetwork = null;
        let cardLast4 = null;

        if (method === 'upi') {
            if (!vpa || !validateVPA(vpa)) {
                return res.status(400)
                    .json({ error: { code: 'INVALID_VPA', description: 'Invalid VPA format' } });
            }
            methodSpecificFields = { vpa };
        } else if (method === 'card') {

            // console.log("Card details", card);
            // console.log("Create Payment with card details", req.body.);
            if (!card) return res.status(400).json({ error: { code: 'BAD_REQUEST_ERROR', description: 'Card details missing' } });

            const { number, expiry_month, expiry_year, cvv, holder_name } = card;

            if (!validateLuhn(number)) {
                return res.status(400).json({ error: { code: 'INVALID_CARD', description: 'Invalid card number' } });
            }

            if (!validateExpiry(expiry_month, expiry_year)) {
                return res.status(400).json({ error: { code: 'EXPIRED_CARD', description: 'Card has expired' } });
            }

            cardNetwork = detectCardNetwork(number);
            cardLast4 = number.slice(-4);
            methodSpecificFields = {
                card_network: cardNetwork,
                card_last4: cardLast4
            };
        } else {
            return res.status(400).json({ error: { code: 'BAD_REQUEST_ERROR', description: 'Invalid payment method' } });
        }

        // 3. Generate ID & Collision Check
        let paymentId = generatePaymentId();
        let retries = 0;
        let unique = false;
        while (!unique && retries < 3) {
            const existing = await query('SELECT id FROM payments WHERE id = $1', [paymentId]);
            if (existing.rows.length === 0) unique = true;
            else {
                paymentId = generatePaymentId();
                retries++;
            }
        }
        if (!unique) throw new Error('Collision Error');

        // 4. Create Payment Record (Pending)
        const created_at = new Date();
        const updated_at = created_at;

        const result = await query(
            `INSERT INTO payments (
                id, merchant_id, order_id, amount, currency, status, method, 
                card_network, card_last4, vpa, created_at, updated_at, captured
            ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, FALSE)
            RETURNING *`,
            [
                paymentId, merchant.id, order.id, order.amount, order.currency,
                'pending', method, cardNetwork, cardLast4, vpa || null,
                created_at, updated_at
            ]
        );
        const payment = result.rows[0];

        // 5. Trigger Background Job
        await paymentQueue.add('process-payment', { paymentId: payment.id });

        // 6. Trigger Webhooks (payment.created, payment.pending)
        const commonPayload = {
            id: payment.id,
            order_id: payment.order_id,
            amount: payment.amount,
            currency: payment.currency,
            method: payment.method,
            status: payment.status,
            created_at: payment.created_at,
            vpa: payment.vpa,
            card_network: payment.card_network,
            card_last4: payment.card_last4
        };

        // Emit payment.created
        await webhookQueue.add('deliver-webhook', {
            merchantId: merchant.id,
            event: 'payment.created',
            payload: {
                event: 'payment.created',
                timestamp: Math.floor(Date.now() / 1000),
                data: { payment: commonPayload }
            }
        });

        // Emit payment.pending
        await webhookQueue.add('deliver-webhook', {
            merchantId: merchant.id,
            event: 'payment.pending',
            payload: {
                event: 'payment.pending',
                timestamp: Math.floor(Date.now() / 1000),
                data: { payment: commonPayload }
            }
        });

        // 6. Return Response
        const responseData: any = {
            id: payment.id,
            order_id: payment.order_id,
            amount: payment.amount,
            currency: payment.currency,
            method: payment.method,
            status: payment.status,
            created_at: payment.created_at,
            updated_at: payment.updated_at
        };

        if (method === 'upi') {
            responseData.vpa = payment.vpa;
        } else {
            responseData.card_network = payment.card_network;
            responseData.card_last4 = payment.card_last4;
        }

        // 7. Store Idempotency Key (24h expiry)
        if (idempotencyKey) {
            const expiresAt = new Date(Date.now() + 24 * 60 * 60 * 1000);
            await query(
                `INSERT INTO idempotency_keys (key, merchant_id, response, expires_at)
                 VALUES ($1, $2, $3, $4)`,
                [idempotencyKey, merchant.id, responseData, expiresAt]
            );
        }

        res.status(201).json(responseData);

    } catch (error) {
        console.error('Create Payment Error:', error);
        res.status(500).json({ error: { code: 'INTERNAL_SERVER_ERROR', description: 'Internal server error' } });
    }
};

export const getPayment = async (req: Request, res: Response) => {
    const { id } = req.params;
    const merchant = req.merchant;

    try {
        const result = await query('SELECT * FROM payments WHERE id = $1 AND merchant_id = $2', [id, merchant.id]);
        if (result.rows.length === 0) {
            return res.status(404).json({ error: { code: 'NOT_FOUND_ERROR', description: 'Payment not found' } });
        }
        res.json(result.rows[0]);
    } catch (error) {
        res.status(500).json({ error: { code: 'INTERNAL_SERVER_ERROR', description: 'Internal Server Error' } });
    }
};

export const capturePayment = async (req: Request, res: Response) => {
    const { id } = req.params;
    const { amount } = req.body;
    const merchant = req.merchant;

    try {
        const result = await query('SELECT * FROM payments WHERE id = $1 AND merchant_id = $2', [id, merchant.id]);

        if (result.rows.length === 0) {
            return res.status(404).json({ error: { code: 'NOT_FOUND_ERROR', description: 'Payment not found' } });
        }

        const payment = result.rows[0];

        // Validation: Verify payment is in capturable state
        if (payment.status !== 'success' || payment.captured === true) {
            return res.status(400).json({
                error: {
                    code: 'BAD_REQUEST_ERROR',
                    description: 'Payment not in capturable state'
                }
            });
        }

        // Update captured = true
        const updateResult = await query(
            'UPDATE payments SET captured = TRUE, updated_at = NOW() WHERE id = $1 RETURNING *',
            [id]
        );
        const updatedPayment = updateResult.rows[0];


        const responseData = {
            ...updatedPayment,
            captured: true
        };

        res.status(200).json(responseData);

    } catch (error) {
        console.error('Capture Error:', error);
        res.status(500).json({ error: { code: 'INTERNAL_SERVER_ERROR', description: 'Internal Server Error' } });
    }
};

export const getPayments = async (req: Request, res: Response) => {
    const merchant = req.merchant;

    try {
        const result = await query(
            `SELECT p.*, 
                    COALESCE((SELECT SUM(amount) FROM refunds r WHERE r.payment_id = p.id AND r.status = 'processed'), 0) as refunded_amount
             FROM payments p 
             WHERE p.merchant_id = $1 
             ORDER BY p.created_at DESC`,
            [merchant.id]
        );
        res.json(result.rows);
    } catch (error) {
        console.error('List Payments Error:', error);
        res.status(500).json({ error: { code: 'INTERNAL_SERVER_ERROR', description: 'Internal Server Error' } });
    }
};
